From abf9bb950f283efd7bb85a5f029615a925a14a51 Mon Sep 17 00:00:00 2001
From: Karthikeyan K <karthikk@habana.ai>
Date: Fri, 08 Oct 2021 19:10:38 +0300
Subject: [PATCH] [SW-56231] sha: padding for checksum descriptor

As the checksum size in image descriptor is adjusted to the
maximum number of bytes of the supported hash alogrithms
(sha384: 48 bytes), add support for padding to map the
indexing between primary and secondary image checksum.

For eg: for platforms which support sha256 as the hashing
algorithm img-ptr consists of checksum of the preboot image
to be loaded, i.e. a sha256(32 byte) hash for each primary
and secondary image consecutively. But now as the boot info
descriptor checksum is of 48 byte, the info descriptor api
look for the checksum at a delta of 48 bytes, hence causing
an indexing mismatch. By adding padding we fix this issue.
These extra padding bytes would not impact the checksum
calculation in the fit validation logic as the hash size is
the criteria to verify the limits of the hash bytes.

Hence this logic can be scaled in case there is a need to
support other hash algorithms with a bigger digest length.

Change-Id: Iec1d3e08c519bf02c9eec6206d728fb9285bef4f
---

diff --git a/board/habanalabs/common/post-image-spi.sh b/board/habanalabs/common/post-image-spi.sh
index 728b420..f449912 100755
--- a/board/habanalabs/common/post-image-spi.sh
+++ b/board/habanalabs/common/post-image-spi.sh
@@ -33,6 +33,19 @@
 
 source ${BOARD_BASE}/post-image-spi.sh
 
+function hl_chksum_padding {
+	# Add zero padding for compatibility with image descriptors size.
+	# Can be scaled for different hash types, independent of platform.
+
+	if [[ ${SHA_CHKSUM} == "sha256sum" ]];then
+		local pad_b=0
+
+		for i in {0..3}; do
+			printf "%08x" $pad_b
+		done
+	fi
+}
+
 function print_image_desc {
 	local img_name=$1
 	local img_num=$2
@@ -82,7 +95,9 @@
 	local preboot_le="${preboot_file}.bin.le"
 	local PRE_CKSUM=`${SHA_CHKSUM} ${preboot_le} | cut -d' ' -f1`
 	echo ${PRE_CKSUM}
+	hl_chksum_padding
 	echo ${PRE_CKSUM}
+	hl_chksum_padding
 }
 
 function prep_image_ptrs {
